src/Authenticator/FormAuthenticator.php

```diff
@@ -100,14 +105,76 @@ protected function _checkLoginUrl(ServerRequestInterface $request)
         $loginUrl = $this->getConfig('loginUrl');
 
         if (!empty($loginUrl)) {
-            if (is_array($loginUrl)) {
-                $loginUrl = Router::url(['_base' => false] + $loginUrl);
-                $this->setConfig('loginUrl', $loginUrl);
+            if (is_string($loginUrl) && $this->_compareStringUrl($loginUrl, $request->getUri()->getPath())) {
+                return true;
             }
 
-            return strcasecmp($request->getUri()->getPath(), $loginUrl) === 0;
+            return $this->_compareArrayUrl($loginUrl, $request);
         }
 
         return true;
     }
+
+    /**
+     * Compare the URL as array
+     *
+     * @param string|array $loginUrl Login URL
+     * @param \Cake\Http\ServerRequestInterface $request Request
+     * @return bool
+     */
+    protected function _compareArrayUrl($loginUrl, ServerRequestInterface $request)
+    {
+        if (!class_exists('\Cake\Routing\Router')
+            || !class_exists('\Cake\Http\ServerRequest')) {
+            return false;
+        }
+
+        $requestUrl = \Cake\Routing\Router::parseRequest($request);
+        unset($request['_matchedRoute']);
+
+        if (is_string($loginUrl)) {
+            $loginUrl = \Cake\Routing\Router::parseRequest((new \Cake\Http\ServerRequest([
+                'uri' => $loginUrl
+            ])));
+
+            unset($loginUrl['_matchedRoute']);
+            $this->setConfig('loginUrl', $loginUrl);
+        }
+
+        $keysToCompare = array_keys($loginUrl);
+        foreach ($keysToCompare as $key) {
+            if (!array_key_exists($key, $requestUrl)
+                || $requestUrl[$key] !== $loginUrl[$key]
+            ) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Checks string URLs for login
+     *
+     * @param string $loginUrl URL of the login to compare against
+     * @param string $requestUrl URL from the request
+     * @return bool
+     */
+    protected function _compareStringUrl($loginUrl, $requestUrl)
+    {
+        if (!is_string($loginUrl)) {
+            throw new InvalidArgumentException();
+        }
+
+        if ($loginUrl === $requestUrl) {
+            return true;
+        }
+
+        $regex = $this->getConfig('loginUrlRegex');
```

Comment:
Having a separate config option is a good idea :+1: Better than using `preg:` prefix as we've done in other places in the past.